### PR TITLE
implement `ensureMove`

### LINKED ISSUE #432: Have compiler warnings for "bug prone use after move" on copyable types
## Motivation

There are situations in which one have a copyable type, but an instance of the type should not be copied implicitly when consumed in a procedure with a sink argument when the instance is also used after the sink. It's possible to accidentally write incorrect code (or less performant code) without having the compiler warning about possible mistakes.

Much like what x-value references argument in C++ are doing: the compiler will force a creation of a x-value to be able to call into the function, without the added benefit of being able to be notified of bugprone use after moves (but tools like clang-tidy could spot instances of these usages, see https://clang.llvm.org/extra/clang-tidy/checks/bugprone-use-after-move.html).

## Description

Consider this snippet:
```nim

type
    X = object
      s: string

proc `=copy`(x: var X, y: X) =
    x.s = "copied " & y.s

proc `=sink`(x: var X, y: X) =
    `=destroy`(x)
    wasMoved(x)
    x.s = "moved " & y.s

proc consume(x: sink X) =
    echo "Consumed: ", x.s
```

Now consider these 3 cases.

### Case 1
```nim
proc main =
    var x = X(s: "abcdefg")
    consume(x)

when isMainModule:
    main()
```
Output will be:
```
Consumed: abcdefg
```
In this case everything is fine, value seems to be moved implicitly (without even reaching a call to =sink, which is even better).

### Case 2
```nim
proc main =
    var x = X(s: "abcdefg")
    consume(x)
    echo "Unconsumed: ", x.s # Let's add this, maybe it was done by mistake

when isMainModule:
    main()
```
Output will be:
```
Consumed: copied abcdefg
Unconsumed: abcdefg
```
This will copy the value `x` when calling `consume` because `x` is accessed after the call as well, so the `sink` parameter is just an optimization, but it's not an enforced specification of how the parameter should be transferred into the procedure. It might be incorrect code, compiler could warn this.

### Case 3
```nim
proc main =
    var x = X(s: "abcdefg")
    consume(move(x))
    echo "Unconsumed: ", x.s

when isMainModule:
    main()
```
Output will be:
```
Consumed: abcdefg
Unconsumed: 
```
This is also suspicious, it might be fine or it might be incorrect code.

So in cases like 2 and 3 it would be great if the compiler could warn about use after move.

### Pragma at procedure definition when sink parameters are defined
```nim
proc consume(x: sink X) {.forcedsink: x.} =
    echo "Consumed: ", x.s
```
Will produce with Cases 2 and 3:
```
Warning: procedure 'consume' requires last read of passed argument 'x' but it seems it used after at line 45; routine: main
```

## Backward incompatibility

None that I can think of in the API.

### PR DESCRIPTION
**Pull request by ringabout**
closes https://github.com/nim-lang/RFCs/issues/432

- [x] better error messages like the reason why it cannot be moved
- [x] add tests

### CODE REVIEW

File: compiler/injectdestructors.nim [hunk 1]
```diff
@@ -1069,6 +1089,11 @@ proc genFieldAccessSideEffects(c: var Con; s: var Scope; dest, ri: PNode; flags:
   result = newTree(nkStmtList, v, snk, c.genWasMoved(newAccess))
 
 proc moveOrCopy(dest, ri: PNode; c: var Con; s: var Scope, flags: set[MoveOrCopyFlag] = {}): PNode =
+  var ri = ri
+  var isEnsureMove = 0
```
Comment (Araq) [inline compiler/injectdestructors.nim:1093]:
```suggestion
  var isEnsureMove = false
```

File: compiler/injectdestructors.nim [hunk 2]
```diff
@@ -1069,6 +1089,11 @@ proc genFieldAccessSideEffects(c: var Con; s: var Scope; dest, ri: PNode; flags:
...
+  if ri.kind in nkCallKinds and ri[0].kind == nkSym and ri[0].sym.magic == mEnsureMove:
+    ri = ri[1]
+    isEnsureMove = 1
```
Comment (Araq) [inline compiler/injectdestructors.nim:1096]:
```suggestion
    isEnsureMove = true
```

**Comment by ringabout** [comment]
```nim
proc isLastRead(n: PNode; c: var Con; s: var Scope): bool =
  if not hasDestructor(c, n.typ): return true
```

Since `isLastRead` always returns true for trivial types, there is no telling whether the symbol can be moved. Should I make `ensureMove` only work for types with destructors?

**Comment by Araq** [comment]
> Since isLastRead always returns true for trivial types, there is no telling whether the symbol can be moved. Should I make ensureMove only work for types with destructors?

No, trivial types are "moved" too. Anything else would not work in a generic context.

**Comment by ringabout** [comment]
```nim
proc foo =
  var x = 1
  let y = ensureMove x # move
  echo (y, x) # (1, 0)
foo()
```

So `x` is moved regardless of whether it is used later?

**Comment by Araq** [comment]
> So x is moved regardless of whether it is used later?

Well yes, or you adapt the code so that the move analyser is run for trivial locations too but that might be too hard.

**Comment by ringabout** [comment]
Sure thing

**Pull request merged**